src/compiler/program.ts

```diff
@@ -1507,6 +1517,97 @@ namespace ts {
             });
         }
 
+        function findLibraryDefinition(searchPath: string) {
+            let typingFilename = "index.d.ts";
+            const packageJsonPath = combinePaths(searchPath, "package.json");
+            if (host.fileExists(packageJsonPath)) {
+                let package: { typings?: string } = {};
+                try {
+                    package = JSON.parse(host.readFile(packageJsonPath));
+                }
+                catch (e) { }
+
+                if (package.typings) {
+                    typingFilename = package.typings;
+                }
+            }
+
+            const combinedPath = normalizePath(combinePaths(searchPath, typingFilename));
+            return host.fileExists(combinedPath) ? combinedPath : undefined;
+        }
+
+        function processReferencedLibraries(file: SourceFile, compilationRoot: string) {
+            const primarySearchPaths = map(getEffectiveLibraryPrimarySearchPaths(), path => combinePaths(compilationRoot, path));
+
+            const failedSearchPaths: string[] = [];
+            const moduleResolutionState: ModuleResolutionState = {
+                compilerOptions: options,
+                host: host,
+                skipTsx: true,
+                traceEnabled: false
+            };
+
+            for (const ref of file.referencedLibraries) {
+                // If we already found this library as a primary reference, or failed to find it, nothing to do
+                const previousResolution = resolvedLibraries[ref.fileName];
+                if (previousResolution && (previousResolution.primary || (previousResolution.resolvedFileName === undefined))) {
+                    continue;
+                }
+
+                let foundIt = false;
+
+                // Check primary library paths
+                for (const primaryPath of primarySearchPaths) {
+                    const searchPath = combinePaths(primaryPath, ref.fileName);
+                    const resolvedFile = findLibraryDefinition(searchPath);
+                    if (resolvedFile) {
+                        processSourceFile(resolvedFile, /*isDefaultLib*/ false, /*isReference*/ true, file, ref.pos, ref.end);
+                        resolvedLibraries[ref.fileName] = { primary: true, resolvedFileName: resolvedFile };

```

Comment:
Shouldn't you add it to `resolvedLibraries` before you call `processSourceFile` on it, in case it (or something it references) has a library reference for the same library? Seems it could get circular otherwise.